### ISSUE TITLE
Master passphrase recovery

### ISSUE THREAD

**Issue by burningtree** [fuxoft/krypta#4]
**What is the safest and not overcomplicated method to create a possibility of recover master passphrase (and related difficulty, salt and indexes)?**

Master passphrase (or a pattern how to assemble it) is safely stored in our brain (memory), but the human body has some disadvantages, there are most important cases:
- death
- brain damage like memory loss

I understand Krypta as a long-term way to keep most-valuable secrets, and therefore using it without some method of recovery can be very unwise. Nobody wants, for example, to irretrievably destroy all his bitcoin savings after death.

The easiest option is to give a master password in a physical form to a trusted person who safely saves it somewhere. But that could be relatively easily broken through social engineering / physical attack, so we need at least to add more entropy.

Here are some my ideas how to achieve it:
- divide passphrase (recovery) into more parts among more people - something like multi-sig
- time factor - passphrase will be automatically released to the trusted person(s) after a specified time of inactivity (for example - last activity on your Google Account, released after 3 months of inactivity)

**Comment by fuxoft** [comment]
This is a complex question that has almost nothing to do with krypta: "How do you safeguard an important secret so that it's not completely lost when you die?"

Any solution to this problem has to have a balance of security and longevity. If you give your passphrase (and other info needed to reconstruct your secrets) to any person, he can steal your passwords or someone else can steal them from them. This is not optimal. If you keep your passphrase only in your head, the passwords are lost when you die. This is also not optimal (if you want your loved ones to inherit your Bitcoins). No solution to this problem is optimal (and cannot be).

As an realistic example, you could give your notes (with Indexes, Difficulty, checksums etc.) to your family members, and put your Passphrase (Passphrase only, not other information) to your lawyer's safe so that your family gets it if you die. Of course, if your family or your lawyer becomes evil, you might have a lot of problems. You can be safer if you give two halves of your Passphrase to two different lawyers. But this has other potential problems, etc. etc...

**Comment by fuxoft** [comment]
The "time factor" solution mentioned by you has many potential problems. If you do this, there has to be a software that releases the Passphrase (when you are not active). That means that this software has to know the Passphrase. Wherever you run this software, it can be hacked (and reveal the Passphrase!) or deleted during a crash... This cannot be used reliably for important passwords.

**Comment by burningtree** [comment]
I know that is a more general problem, but every (serious) Krypta user doing some kind of preparation for recovery, so some recommendation or talk about it would make sense.

Another dimension of the problem is technical skills which are required to operate the Krypta - making recovery much more difficult. Operate command-line and know how to run LuaJIT script is not a standard ability of today's family members. I believe that to have some "recovery" options will be very helpful, maybe some special recovery web, GUI or some pre-build ISO with ELI5 instructions. It doesn't make sense when we all recreate a wheel again.

I write a script, which builds [minimal bootable FreeBSD with Krypta](https://github.com/burningtree/krypta-liveos) for i386 / AMD64 architecture. It runs completely in memory and its based on [mfsBSD](http://mfsbsd.vx.sk/).

> The "time factor" solution mentioned by you has many potential problems. If you do this, there has to be a software that releases the Passphrase (when you are not active). That means that this software has to know the Passphrase. Wherever you run this software, it can be hacked (and reveal the Passphrase!) or deleted during a crash... This cannot be used reliably for important passwords.

Yes, I agree that to use *only* this approach is very dangerous. But as an extension to multi-person, it can be in some circumstances very helpful against social engineering at least. I imagine something more complex like time-factor application(s) can release half of the Passphrase, and the second half can hold trusted persons. Its possible to make redundant mirrors or divide it into more sources, which releases independently. But yeah, it's just hypothesis, a user need really reliable and trustworthy hardware, with regular maintenance, if he wants to run it himself. It's a good case for some smart-contract solution.
